app/src/main/java/io/islnd/android/islnd/messaging/Rest.java

```diff
@@ -244,4 +245,75 @@ public static void postComment(EncryptedComment encryptedComment, String apiKey)
 
         return null;
     }
+
+    public static long getServerTimeOffsetMillis(int repetitions, String apiKey) throws IOException {
+        // account for network delay by using OkHttpClient's Interceptor
+        class DelayInterceptor implements okhttp3.Interceptor {
+            private long requestTimeMillis;
+            private long networkDelayNanos;
+            public long getRequestTimeMillis() {
+                return requestTimeMillis;
+            }
+            public long getNetworkDelayNanos() {
+                return networkDelayNanos;
+            }
+
+            @Override
+            public okhttp3.Response intercept(Chain chain) throws IOException {
+                networkDelayNanos = 0;
+                requestTimeMillis = System.currentTimeMillis();
+
+                // perform request, and get delay
+                long t1 = System.nanoTime();
+                okhttp3.Response response = chain.proceed(chain.request());
+                long t2 = System.nanoTime();
+
+                // assume half of round trip
+                networkDelayNanos = (t2 - t1) / 2;
+
+                return response;
+            }
+        }
+        DelayInterceptor delayInterceptor = new DelayInterceptor();
+        okhttp3.OkHttpClient client = new okhttp3.OkHttpClient.Builder()
+                .addNetworkInterceptor(delayInterceptor)
+                .build();
+        Retrofit retrofit = new Retrofit.Builder()
+                .baseUrl(HOST)
+                .addConverterFactory(GsonConverterFactory.create())
+                .client(client)
+                .build();
+
+        RestInterface service = retrofit.create(RestInterface.class);
+
+        // record lowest delay for all attempts
+        long minNetworkDelayNanos = Long.MAX_VALUE;
+        long mostAccurateOffsetMillis = 0;
+
+        for (int i = 0; i < repetitions; i++) {
```

Comment:
I am going back and forth on this, so I will leave it up to you. Until now, all of the Rest methods were wrappers for a single rest call. This changes that pattern by having multiple calls and some logic and delays and stuff like that. My instinct would be to put that in the message layer or some other class and keep the methods Rest class simple. I guess the advantage here is that we only create the client and the retrofit instance once, so there is less overhead there. I think that overhead is not significant compared to the network delay, but I could go either way. Go with whatever seems best to you.